core/genesis.go

```diff
@@ -589,33 +593,50 @@ func (g *Genesis) ToBlock() *types.Block {
 
 // toBlockWithRoot constructs the genesis block with the given genesis state root.
 func (g *Genesis) toBlockWithRoot(stateRoot, storageRootMessagePasser common.Hash) *types.Block {
-	head := &types.Header{
-		Number:     new(big.Int).SetUint64(g.Number),
-		Nonce:      types.EncodeNonce(g.Nonce),
-		Time:       g.Timestamp,
-		ParentHash: g.ParentHash,
-		Extra:      g.ExtraData,
-		GasLimit:   g.GasLimit,
-		GasUsed:    g.GasUsed,
-		BaseFee:    g.BaseFee,
-		Difficulty: g.Difficulty,
-		MixDigest:  g.Mixhash,
-		Coinbase:   g.Coinbase,
-		Root:       stateRoot,
-	}
-	if g.GasLimit == 0 {
-		head.GasLimit = params.GenesisGasLimit
-	}
-	if g.Difficulty == nil && g.Mixhash == (common.Hash{}) {
-		head.Difficulty = params.GenesisDifficulty
+	number := new(big.Int).SetUint64(g.Number)
+	head := &types.Header{}
+	// If we are in an initGenesis flow and this is a pre-gingerbread block, then
+	// set the block to be pre-gingerbread. This affects the way it is encoded,
+	// which affects its hash. Ideally we could do without the InitingGenesis
+	// condition, but TestFeeHistory, makes use of chains with pre-gingerbread
+	// genesis blocks without actually accounting for the different structure of
+	// the pre-gingerbread block.
+	if g.InitingGenesis() && g.Config.GingerbreadBlock != nil && !g.Config.IsGingerbread(number) {
+		head = types.NewPreGingerbreadHeader()
+	}
+
+	head.Number = number
+	head.Nonce = types.EncodeNonce(g.Nonce)
+	head.Time = g.Timestamp
+	head.ParentHash = g.ParentHash
+	head.Extra = g.ExtraData
+	head.GasLimit = g.GasLimit
+	head.GasUsed = g.GasUsed
+	head.BaseFee = g.BaseFee
+	head.Difficulty = g.Difficulty
+	head.MixDigest = g.Mixhash
+	head.Coinbase = g.Coinbase
+	head.Root = stateRoot
+
+	// When initialising the genesis we don't want to override unset gasLimit or
+	// difficulty fields since these fields were not present at the celo chain genesis.
+	if !g.Config.IsMigratedChain() {
+		if g.GasLimit == 0 {
+			head.GasLimit = params.GenesisGasLimit
+		}
+		if g.Difficulty == nil && g.Mixhash == (common.Hash{}) {
+			head.Difficulty = params.GenesisDifficulty
+		}
 	} else if g.Difficulty == nil {
 		// In the case of migrated chains we ensure a zero rather than nil difficulty.
 		head.Difficulty = new(big.Int)
 	}
+
 	if g.Config != nil && g.Config.IsLondon(common.Big0) {
 		if g.BaseFee != nil {
 			head.BaseFee = g.BaseFee
 		} else {
+			// If defaults are not ignored, set default values for base fee.
```

Comment:
Just to clarify my own understanding, how do we know defaults are not ignored at this point in the code?